File: CSC455-Game-Programming/ShootThemCircles/core/src/com/jeff/game/models/systems/VelocitySystem.java
package com.jeff.game.models.systems;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.ashley.systems.IteratingSystem;
import com.jeff.game.models.components.TransformComponent;
import com.jeff.game.models.components.VelocityComponent;

/**
 * Velocity system.
 */
public class VelocitySystem extends IteratingSystem {

    private final ComponentMapper<TransformComponent> tm = ComponentMapper.getFor(TransformComponent.class);
    private final ComponentMapper<VelocityComponent> vm = ComponentMapper.getFor(VelocityComponent.class);

    public VelocitySystem() {
        super(Family.all(TransformComponent.class, VelocityComponent.class).get());
    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {
        final TransformComponent t = tm.get(entity);
        final VelocityComponent v = vm.get(entity);
        t.pX += (v.x * deltaTime);
        t.pY += (v.y * deltaTime);
    }
}

File: CSC455-Game-Programming/Chaser/core/src/com/jeff/chaser/models/systems/VelocitySystem.scala
package com.jeff.chaser.models.systems

import com.badlogic.ashley.core.{ComponentMapper, Entity, Family}
import com.badlogic.ashley.systems.IteratingSystem
import com.badlogic.gdx.math.MathUtils
import com.jeff.chaser.models.components.{TransformComponent, VelocityComponent}


class VelocitySystem extends IteratingSystem(Family.all(classOf[VelocityComponent], classOf[TransformComponent]).get()) {

  private val vm = ComponentMapper.getFor(classOf[VelocityComponent])
  private val tm = ComponentMapper.getFor(classOf[TransformComponent])

  override def processEntity(entity: Entity, deltaTime: Float): Unit = {
    val trans = tm.get(entity)
    val vel = vm.get(entity)
    trans.x += (clamp(vel.x, vel.maxX) * deltaTime)
    trans.y += (clamp(vel.y, vel.maxY) * deltaTime)
    vel.targetX = clamp(vel.targetX, vel.maxX)
    vel.targetY = clamp(vel.targetY, vel.maxY)
  }

  private def clamp(value: Float, max: Float): Float = {
    MathUtils.clamp(value, -max, max)
  }

}
